File: crates/core/src/tracker.rs
//! The IO-free core: feed it lines, it hands back events, state and a cache

use serde::{Deserialize, Serialize};

use crate::types::{Sighting, State, Update, Way, Wire};

/// With no heartbeat for this many seconds the connection counts as dead
/// (the server beats roughly every 10 s)
pub const HEARTBEAT_TIMEOUT: i64 = 30;

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct Tracker {
  /// Last known crossing, which doubles as the status cache
  pub last: Option<Sighting>,
  pub last_heartbeat: Option<i64>,
  pub count_in: u32,
  pub count_out: u32,
  /// Unfinished line carried between stream chunks
  #[serde(skip)]
  buffer: String,
}

impl Tracker {
  #[must_use]
  pub fn new() -> Self {
    Self::default()
  }

  /// Feeds one SSE line or a bare JSON object
  ///
  /// Returns `None` for comments, `event:`/`id:` fields and garbage.
  pub fn feed_line(&mut self, line: &str) -> Option<Update> {
    let line = line.trim_end_matches(['\r', '\n']).trim();
    let payload = match line.strip_prefix("data:") {
      Some(rest) => rest.trim_start(),
      None if line.starts_with('{') => line,
      None => return None,
    };
    let wire: Wire = serde_json::from_str(payload).ok()?;
    Some(self.apply(wire))
  }

  /// Feeds any slice of the stream, including one cut mid-line
  pub fn feed_chunk(&mut self, chunk: &str) -> Vec<Update> {
    self.buffer.push_str(chunk);
    let mut updates = Vec::new();
    while let Some(newline) = self.buffer.find('\n') {
      let line: String = self.buffer.drain(..=newline).collect();
      if let Some(update) = self.feed_line(&line) {
        updates.push(update);
      }
    }
    // ponytail: the buffer only grows to the end of a line; a stream without
    // '\n' would run out of memory first, and such a server is broken anyway.
    updates
  }

  /// Feeds the `/rest/v1/last-status` response, which has the same shape as an SSE frame
  pub fn feed_rest(&mut self, json: &str) -> Option<Update> {
    let wire: Wire = serde_json::from_str(json).ok()?;
    Some(self.apply(wire))
  }

  fn apply(&mut self, wire: Wire) -> Update {
    match wire {
      Wire::Heartbeat { time } => {
        self.last_heartbeat = Some(time.max(self.last_heartbeat.unwrap_or(i64::MIN)));
        Update::Heartbeat { time }
      }
      Wire::Pepito { way, time, img } => {
        if let Some(previous) = &self.last
          && previous.time == time
          && previous.way == way
        {
          return Update::Duplicate { time };
        }
        let changed = self
          .last
          .as_ref()
          .is_none_or(|previous| previous.way != way);
        match way {
          Way::In => self.count_in += 1,
          Way::Out => self.count_out += 1,
        }
        let sighting = Sighting {
          way,
          time,
          img,
          text: None,
        };
        // An event older than the cache (a late reconnect) still counts, but it
        // must not roll the state backwards.
        if self
          .last
          .as_ref()
          .is_none_or(|previous| time >= previous.time)
        {
          self.last = Some(sighting.clone());
        }
        Update::Sighting { sighting, changed }
      }
    }
  }

  /// The whole state in one object: where the cat is, how stale the cache is,
  /// whether the heartbeat is alive
  ///
  /// The shape of this JSON is the public contract of every binding, which is
  /// why it is assembled here once instead of separately in each of them.
  #[must_use]
  pub fn status(&self, now: i64) -> Status {
    Status {
      state: self.state().as_str(),
      since: self.state().since(),
      last: self.last.clone(),
      age: self.age(now),
      heartbeat_age: self.heartbeat_age(now),
      healthy: self.healthy(now),
      count_in: self.count_in,
      count_out: self.count_out,
    }
  }

  #[must_use]
  pub fn state(&self) -> State {
    match &self.last {
      Some(sighting) if sighting.way == Way::In => State::Home {
        since: sighting.time,
      },
      Some(sighting) => State::Away {
        since: sighting.time,
      },
      None => State::Unknown,
    }
  }

  /// Age of the cache in seconds
  #[must_use]
  pub fn age(&self, now: i64) -> Option<i64> {
    self.last.as_ref().map(|sighting| now - sighting.time)
  }

  /// Whether a REST refresh is worth it
  #[must_use]
  pub fn is_stale(&self, now: i64, max_age: i64) -> bool {
    self.age(now).is_none_or(|age| age > max_age)
  }

  #[must_use]
  pub fn heartbeat_age(&self, now: i64) -> Option<i64> {
    self.last_heartbeat.map(|beat| now - beat)
  }

  /// Whether the stream is alive. No heartbeat means it is time to reconnect.
  #[must_use]
  pub fn healthy(&self, now: i64) -> bool {
    self
      .heartbeat_age(now)
      .is_some_and(|age| age <= HEARTBEAT_TIMEOUT)
  }

  /// Cache to persist between processes, which matters for PHP where every
  /// request starts from nothing
  #[must_use]
  pub fn snapshot(&self) -> String {
    serde_json::to_string(self).unwrap_or_else(|_| "{}".into())
  }

  #[must_use]
  pub fn restore(json: &str) -> Tracker {
    serde_json::from_str(json).unwrap_or_default()
  }
}

/// The tracker's state captured at one instant. See [`Tracker::status`].
#[derive(Debug, Clone, Serialize)]
pub struct Status {
  pub state: &'static str,
  pub since: Option<i64>,
  pub last: Option<Sighting>,
  pub age: Option<i64>,
  pub heartbeat_age: Option<i64>,
  pub healthy: bool,
  pub count_in: u32,
  pub count_out: u32,
}

File: php/src/Tracker.php
<?php

declare(strict_types=1);

namespace Pepito;

/**
 * State, cache, dedup and heartbeat in plain PHP, the default backend of Client.
 *
 * @internal Use Client, which picks the backend.
 */
final class Tracker
{
    /** With no heartbeat for this many seconds the connection counts as dead (the server beats roughly every 10 s). */
    public const HEARTBEAT_TIMEOUT = 30;

    private const U32_MAX = 4294967295;

    /** @var array{way: string, time: int, img?: string, text?: string}|null */
    private ?array $last = null;

    private ?int $lastHeartbeat = null;

    private int $countIn = 0;

    private int $countOut = 0;

    /** Unfinished line carried between stream chunks. */
    private string $buffer = '';

    /** A snapshot that does not parse means a fresh tracker. */
    public function __construct(?string $snapshot = null)
    {
        $raw = $snapshot === null ? null : json_decode($snapshot, true);
        if (! \is_array($raw) || ! self::isCount($raw['count_in'] ?? null) || ! self::isCount($raw['count_out'] ?? null)) {
            return;
        }
        $last = isset($raw['last']) ? self::restoreSighting($raw['last']) : null;
        $beat = $raw['last_heartbeat'] ?? null;
        if ((isset($raw['last']) && $last === null) || ($beat !== null && ! \is_int($beat))) {
            return;
        }
        $this->last = $last;
        $this->lastHeartbeat = $beat;
        $this->countIn = $raw['count_in'];
        $this->countOut = $raw['count_out'];
    }

    /**
     * Feeds any slice of the SSE stream, including one cut mid-line.
     *
     * @return list<array<string, mixed>>
     */
    public function feed(string $chunk): array
    {
        $this->buffer .= $chunk;
        $updates = [];
        while (($newline = strpos($this->buffer, "\n")) !== false) {
            $line = trim(substr($this->buffer, 0, $newline));
            $this->buffer = substr($this->buffer, $newline + 1);
            if (str_starts_with($line, 'data:')) {
                $update = $this->feedRest(ltrim(substr($line, 5)));
            } elseif (str_starts_with($line, '{')) {
                $update = $this->feedRest($line);
            } else {
                continue;
            }
            if ($update !== null) {
                $updates[] = $update;
            }
        }

        return $updates;
    }

    /**
     * Feeds the REST last-status response, which has the same shape as an SSE frame.
     *
     * @return array<string, mixed>|null
     */
    public function feedRest(string $json): ?array
    {
        $raw = json_decode($json, true);
        if (! \is_array($raw) || ! \is_int($raw['time'] ?? null)) {
            return null;
        }
        $time = $raw['time'];
        $event = $raw['event'] ?? null;
        if ($event === 'heartbeat') {
            $this->lastHeartbeat = max($time, $this->lastHeartbeat ?? PHP_INT_MIN);

            return ['kind' => 'heartbeat', 'time' => $time];
        }
        $way = $raw['type'] ?? null;
        $img = $raw['img'] ?? null;
        if ($event !== 'pepito' || ! self::isWay($way) || ($img !== null && ! \is_string($img))) {
            return null;
        }

        $previous = $this->last;
        if ($previous !== null && $previous['time'] === $time && $previous['way'] === $way) {
            return ['kind' => 'duplicate', 'time' => $time];
        }
        $changed = $previous === null || $previous['way'] !== $way;
        if ($way === 'in') {
            $this->countIn++;
        } else {
            $this->countOut++;
        }
        $sighting = self::sighting($way, $time, $img);
        // An event older than the cache (a late reconnect) still counts, but it
        // must not roll the state backwards.
        if ($previous === null || $time >= $previous['time']) {
            $this->last = $sighting;
        }

        return ['kind' => 'sighting', ...$sighting, 'changed' => $changed];
    }

    /**
     * Where the cat is, how stale the cache is, whether the heartbeat is alive.
     *
     * @return array<string, mixed>
     */
    public function status(int $now): array
    {
        $last = $this->last;
        $heartbeatAge = $this->lastHeartbeat === null ? null : $now - $this->lastHeartbeat;
        $state = 'unknown';
        if ($last !== null) {
            $state = $last['way'] === 'in' ? 'home' : 'away';
        }

        return [
            'state' => $state,
            'since' => $last['time'] ?? null,
            'last' => $last,
            'age' => $last === null ? null : $now - $last['time'],
            'heartbeat_age' => $heartbeatAge,
            'healthy' => $heartbeatAge !== null && $heartbeatAge <= self::HEARTBEAT_TIMEOUT,
            'count_in' => $this->countIn,
            'count_out' => $this->countOut,
        ];
    }

    /** The same JSON as the FFI backend writes, so either one restores it. */
    public function snapshot(): string
    {
        return json_encode([
            'last' => $this->last,
            'last_heartbeat' => $this->lastHeartbeat,
            'count_in' => $this->countIn,
            'count_out' => $this->countOut,
        ], JSON_THROW_ON_ERROR | JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE);
    }

    /**
     * No image or text means no key, not a null one.
     *
     * @return array{way: string, time: int, img?: string, text?: string}
     */
    public static function sighting(string $way, int $time, ?string $img = null, ?string $text = null): array
    {
        return array_filter(
            ['way' => $way, 'time' => $time, 'img' => $img, 'text' => $text],
            static fn (mixed $value): bool => $value !== null,
        );
    }

    public static function isWay(mixed $way): bool
    {
        return $way === 'in' || $way === 'out';
    }

    private static function restoreSighting(mixed $raw): ?array
    {
        if (! \is_array($raw) || ! self::isWay($raw['way'] ?? null) || ! \is_int($raw['time'] ?? null)) {
            return null;
        }
        $img = $raw['img'] ?? null;
        $text = $raw['text'] ?? null;
        if (($img !== null && ! \is_string($img)) || ($text !== null && ! \is_string($text))) {
            return null;
        }

        return self::sighting($raw['way'], $raw['time'], $img, $text);
    }

    private static function isCount(mixed $value): bool
    {
        return \is_int($value) && $value >= 0 && $value <= self::U32_MAX;
    }
}
